services/util/test/TestSesameWindowed.cpp

```diff
@@ -0,0 +1,523 @@
+            {
+                SendMessageStreamAvailableWithWriter(infra::ConstructBin().Value<uint8_t>(3).Value<infra::LittleEndian<uint16_t>>(releasedSize).Vector());
+            }));
+    }
+
+    void ExpectRequestSendMessageForMessage(uint16_t size, const std::vector<uint8_t>& expected)
+    {
+        EXPECT_CALL(base, RequestSendMessage(size)).WillOnce(testing::Invoke([this, expected](uint16_t size)
+            {
+                SendMessageStreamAvailableWithWriter(infra::ConstructBin().Value<uint8_t>(4)(expected).Vector());
+            }));
+    }
+
+    void ExpectSendMessageStreamAvailable(const std::vector<uint8_t>& data)
+    {
+        EXPECT_CALL(observer, SendMessageStreamAvailable(testing::_)).WillOnce(testing::Invoke([data](infra::SharedPtr<infra::StreamWriter>&& writer)
+            {
+                infra::DataOutputStream::WithErrorPolicy stream(*writer);
+                stream << infra::MakeRange(data);
+            }));
+    }
+
+    void ExpectSendMessageStreamAvailableAndSaveWriter()
+    {
+        EXPECT_CALL(observer, SendMessageStreamAvailable(testing::_)).WillOnce(testing::SaveArg<0>(&savedWriter));
+    }
+
+    void ExpectReceivedMessage(const std::string& expected)
+    {
+        EXPECT_CALL(observer, ReceivedMessage(testing::_)).WillOnce(testing::Invoke([expected](infra::SharedPtr<infra::StreamReaderWithRewinding>&& reader)
+            {
+                infra::DataInputStream::WithErrorPolicy stream(*reader);
+                std::string text(stream.Available(), 0);
+                stream >> infra::ByteRange(reinterpret_cast<uint8_t*>(text.data()), reinterpret_cast<uint8_t*>(text.data() + text.size()));
+
+                EXPECT_EQ(expected, text);
+            }));
+    }
+
+    void ExpectReceivedMessageAndSaveReader(const std::string& expected)
+    {
+        EXPECT_CALL(observer, ReceivedMessage(testing::_)).WillOnce(testing::SaveArg<0>(&savedReader));
+    }
+
+    testing::StrictMock<services::SesameEncodedMock> base;
+    std::vector<uint8_t> expectedMessage;
+    std::vector<uint8_t> sentData;
+    infra::NotifyingSharedOptional<infra::StdVectorOutputStreamWriter> writer;
+    infra::Execute execute{ [this]()
+        {
+            EXPECT_CALL(base, MaxSendMessageSize()).WillOnce(testing::Return(16));
+            EXPECT_CALL(base, MessageSize(3)).WillOnce(testing::Return(5));
+            ExpectRequestSendMessageForInit(16);
+        } };
+    infra::SharedOptional<infra::StdVectorInputStreamReader::WithStorage> reader;
+    services::SesameWindowed communication{ base };
+    testing::StrictMock<services::SesameObserverMock> observer{ communication };
+    infra::SharedPtr<infra::StreamWriter> savedWriter;
+    infra::SharedPtr<infra::StreamReaderWithRewinding> savedReader;
+};
+
+TEST_F(SesameWindowedTest, MaxSendMessageSize)
+{
+    // When the cobs layer is able to send a 16 byte message, then 18 bytes (cobs start plus delimiter) are available
```

Comment:
I'm quite surprised to see information about COBS in Windowing class, that in my mind should have been irrelevant for this class...

Reply:
SesameWindowed asks the layer below it how long a message is going to be via MessageSize(...)